Src/Plugins/Security/SEC101_055.CmdkeyValidator.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Linq;
+
+using Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security.Utilities;
+using Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk;
+using Microsoft.RE2.Managed;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security
+{
+    public class CmdkeyValidator : ValidatorBase
+    {
+        internal static CmdkeyValidator Instance;
+        internal static IRegex RegexEngine;
+
+        private const string UserRegex = @"\/user:\s*([^\[""'\s`%]{2,})";
+
+        static CmdkeyValidator()
+        {
+            Instance = new CmdkeyValidator();
+            RegexEngine = RE2Regex.Instance;
+        }
+
+        public static string IsValidStatic(ref string matchedPattern,
+                                           ref Dictionary<string, string> groups,
+                                           ref string failureLevel,
+                                           ref string fingerprint,
+                                           ref string message)
+        {
+            return IsValidStatic(Instance,
+                                 ref matchedPattern,
+                                 ref groups,
+                                 ref failureLevel,
+                                 ref fingerprint,
+                                 ref message);
+        }
+
+        protected override string IsValidStaticHelper(ref string matchedPattern,
+                                                      ref Dictionary<string, string> groups,
+                                                      ref string failureLevel,
+                                                      ref string fingerprintText,
+                                                      ref string message)
+        {
+            if (!groups.TryGetNonEmptyValue("password", out string password))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            if (IsInvalidPassword(password))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            string account = ParseExpression(RegexEngine, matchedPattern, UserRegex);
+            account = account?.Replace("/user:", string.Empty).Trim();
+
+            fingerprintText = new Fingerprint()
+            {
+                Account = account,
+                Password = password,
+            }.ToString();
+
+            // Manual use of the regex against CodeAsData produces highly confident results,
+            // so return AuthorizedWarning rather than Unknown.
+            return nameof(ValidationState.AuthorizedWarning);
+        }
+
+        private static bool IsInvalidPassword(string potentialPassword)
```

Comment:
>IsInvalidPassword [](start = 28, length = 17)

Have we got this right? Which is clearer:

IsValidPassword and !IsValidPassword

or 
IsInvalidPassord and !InvalidPassword

my open question to you.

Reply:
I sincerely had this debate with myself moments before checking in because I actually reversed the boolean logic by mistake.  I chose to leave the method as "IsInvalid" because I can't say with confidence that a password that passes every check within is "valid".  There may be patterns which clearly identify an invalid passwords that I have not implemented.  I can say for sure that passwords failing one of these conditions is almost certainly Invalid.  I can't say that passwords passing all these conditions is valid.

---
In reply to: [598012574](https://github.com/microsoft/sarif-pattern-matcher/pull/317#discussion_r598012574) [](ancestors = 598012574)